test/lein_monolith/dependency_test.clj

```diff
@@ -70,11 +73,148 @@
+
+(defn check-cycle-error [deps smlest-cycles]
+  (doseq [deps (maps-like 10 deps)]
+    (let [^Exception e (try (dep/topological-sort deps)
+                            (catch Exception e e))]
+      (is (instance? IExceptionInfo e)
+          (str "Didn't throw an exception\ndeps: " deps))
+      (is (re-find #"Dependency cycles? detected" (.getMessage e)))
+      ; pretty printed dependency cycle appears in msg
+      #_
+      (is (some (fn [c]
+                  (re-find (re-pattern
+                             (apply str (str/join " -> " c)))
+                           (.getMessage e)))
+                smlest-cycles)
+          (str "Didn't print out smallest cycle:\n"
+               "deps: " deps "\n"
+               "smallest-cycles: " smlest-cycles "\n"
+               "actual message: " (.getMessage e)))
+      (is (->> e ex-data :cycles (some smlest-cycles))
+          (str "Didn't include smallest cycle:\n"
+               "deps: " deps "\n"
+               "smallest-cycles: " smlest-cycles "\n"
+               "actual cycles: " (->> e ex-data :cycles) "\n"
+               "actual message: " (.getMessage e))))))
 
 (deftest topological-sorting
   (let [deps {:a #{}, :b #{:a}, :c #{:a :b} :x #{:b} :y #{:c}}]
     (is (= [:a :b :c :x :y] (dep/topological-sort deps)))
     (is (= [:b :c :x] (dep/topological-sort deps [:x :c :b]))))
-  (let [deps {:a #{:b}, :b #{:c}, :c #{:a}}]
-    (is (thrown-with-msg? Exception #"cycle detected"
-          (dep/topological-sort deps)))))
+  (check-cycle-error {:a #{:b}, :b #{:c}, :c #{:a}}
+                     #{[:a :b :c :a]
+                       [:b :c :a :b]
+                       [:c :a :b :c]})
+  (doseq [size [5 10]] ;gen cycles of these sizes (higher is very slow)
+    (let [[deps smallest-cycles] (gen-dep-cycle size)]
+      (doseq [c (maps-like 5 deps)] ;shuffle deps order <..> times
+        (check-cycle-error c smallest-cycles)))))
+
+(comment
+  (println
+    (dep/topological-sort
+      {:a #{:b},
+       :b #{:c :a},
+       :c #{:d},
+       :d #{:a :b}}))
+  )
+
+(deftest pretty-cycle-test
+  (let [cycle+pretty
+        {[1 1]
+         ["+ 1"
+          "^\\"
+          "|_|"]
+
+         [1 2 1]
+         ["+ 1"
+          "^ + 2"
+          "|_/"]
+
+         [1 2 3 1]
+         ["+ 1"
+          "^ + 2"
+          "|  + 3"
+          "|_/"]
+
+         [1 2 3 4 1]
+         ["+ 1"
+          "^ + 2"
+          "|  + 3"
+          "|   + 4"
+          "|__/"]
+         
+         [1 2 3 4 5 1]
+         ["+ 1"
+          "^ + 2"
+          "|  + 3"
+          "|   + 4"
+          "|    + 5"
+          "|___/"]}]
+    (every? (fn [[c strs]]
+              (let [expected (str/join \newline strs)
+                    actual (dep/pretty-cycle c)]
+                (is (= expected actual)
+                    (str "Pretty representation of cycle " c ":\n"
+                         "Expected:\n" expected "\n\n"
+                         "Actual:\n" actual))))
```

Comment:
The assertion report is already going to include the expected and actual values, you don't need to put them in the message here.

Reply:
Changed to a `testing` wrapper.